crates/corro-agent/src/agent/util.rs

```diff
@@ -894,23 +942,100 @@ pub async fn process_multiple_changes(
                 version: None,
             })?;
 
+        let mut last_cleared: Option<Timestamp> = None;
         for (actor_id, versions_set) in overwritten {
+            if actor_id != agent.actor_id() {
+                warn!("clearing and setting timestamp for empties from a different node");
+            }
             for versions in versions_set {
-                store_empty_changeset(&tx, actor_id, versions)?;
+                let ts = Timestamp::from(agent.clock().new_timestamp());
+                let inserted = store_empty_changeset(&tx, actor_id, versions, ts)?;
+                if inserted > 0 {
+                    last_cleared = Some(ts);
+                }
             }
         }
 
+        for (actor_id, ts_versions) in &sent_empties {
+            println!("handling empties in sent_empties");
+
+            let mut max_ts = Default::default();
+            for (versions, ts) in ts_versions.iter() {
+                for version in versions {
+                    store_empty_changeset(&tx, *actor_id, version.clone(), *ts)?;
+                    if *ts > max_ts {
+                        max_ts = *ts;
+                    }
+                }
+            }
+
+            let mut snap = match snapshots.remove(&actor_id) {
+                Some(snap) => snap,
+                None => {
+                    let booked = {
+                        bookie
+                            .blocking_write(format!(
+                                "process_multiple_changes(for_actor_blocking):{actor_id}",
+                            ))
+                            .ensure(*actor_id)
+                    };
+
+                    let booked_write = booked.blocking_write(format!(
+                        "process_multiple_changes(booked writer, handling requested empties):{actor_id}",
+                    ));
+                    booked_write.snapshot()
+                }
+            };
+
+            snap.update_cleared_ts(&tx, max_ts)
```

Comment:
same comment as before. we only want to store that if it's during sync with the originating node

Reply:
It only sends a `SyncNeed::Empty` to a node's that its sync'ing with and each of those nodes sends their empties. But you are right, will add a check when it receives the empty changeset in `parallel_sync`

Reply:
added the check!